Mathlib/LinearAlgebra/Matrix/ToLin.lean

```diff
@@ -0,0 +1,1001 @@
+    M.mulVec (LinearMap.stdBasis R (fun _ => R) j 1) = Mᵀ j :=
+  funext fun i => Matrix.mulVec_stdBasis M i j
+#align matrix.mul_vec_std_basis_apply Matrix.mulVec_stdBasis_apply
+
+theorem Matrix.range_mulVecLin (M : Matrix m n R) :
+    LinearMap.range M.mulVecLin = span R (range Mᵀ) := by
+  letI := Classical.decEq n
+  simp_rw [range_eq_map, ← supᵢ_range_stdBasis, Submodule.map_supᵢ, range_eq_map, ←
+    Ideal.span_singleton_one, Ideal.span, Submodule.map_span, image_image, image_singleton,
+    Matrix.mulVecLin_apply, M.mulVec_stdBasis_apply, supᵢ_span, range_eq_unionᵢ]
+#align matrix.range_mul_vec_lin Matrix.range_mulVecLin
+
+variable [DecidableEq n]
+
+/-- Linear maps `(n → R) →ₗ[R] (m → R)` are linearly equivalent to `Matrix m n R`. -/
+def LinearMap.toMatrix' : ((n → R) →ₗ[R] m → R) ≃ₗ[R] Matrix m n R where
+  toFun f := of fun i j => f (stdBasis R (fun _ => R) j 1) i
+  invFun := Matrix.mulVecLin
+  right_inv M := by
+    ext (i j)
+    simp only [Matrix.mulVec_stdBasis, Matrix.mulVecLin_apply, of_apply]
+  left_inv f := by
+    apply (Pi.basisFun R n).ext
+    intro j; ext i
+    simp only [Pi.basisFun_apply, Matrix.mulVec_stdBasis, Matrix.mulVecLin_apply, of_apply]
+  map_add' f g := by
+    ext (i j)
+    simp only [Pi.add_apply, LinearMap.add_apply, of_apply, Matrix.add_apply]
+  map_smul' c f := by
+    ext (i j)
+    simp only [Pi.smul_apply, LinearMap.smul_apply, RingHom.id_apply, of_apply, Matrix.smul_apply]
+#align linear_map.to_matrix' LinearMap.toMatrix'
+
+/-- A `Matrix m n R` is linearly equivalent to a linear map `(n → R) →ₗ[R] (m → R)`.
+
+Note that the forward-direction does not require `DecidableEq` and is `Matrix.vecMulLin`. -/
+def Matrix.toLin' : Matrix m n R ≃ₗ[R] (n → R) →ₗ[R] m → R :=
+  LinearMap.toMatrix'.symm
+#align matrix.to_lin' Matrix.toLin'
+
+theorem Matrix.toLin'_apply' (M : Matrix m n R) : Matrix.toLin' M = M.mulVecLin :=
+  rfl
+#align matrix.to_lin'_apply' Matrix.toLin'_apply'
+
+@[simp]
+theorem LinearMap.toMatrix'_symm :
+    (LinearMap.toMatrix'.symm : Matrix m n R ≃ₗ[R] _) = Matrix.toLin' :=
+  rfl
+#align linear_map.to_matrix'_symm LinearMap.toMatrix'_symm
+
+@[simp]
+theorem Matrix.toLin'_symm :
+    (Matrix.toLin'.symm : ((n → R) →ₗ[R] m → R) ≃ₗ[R] _) = LinearMap.toMatrix' :=
+  rfl
+#align matrix.to_lin'_symm Matrix.toLin'_symm
+
+@[simp]
+theorem LinearMap.toMatrix'_toLin' (M : Matrix m n R) : LinearMap.toMatrix' (Matrix.toLin' M) = M :=
+  LinearMap.toMatrix'.apply_symm_apply M
+#align linear_map.to_matrix'_to_lin' LinearMap.toMatrix'_toLin'
+
+@[simp]
+theorem Matrix.toLin'_toMatrix' (f : (n → R) →ₗ[R] m → R) :
+    Matrix.toLin' (LinearMap.toMatrix' f) = f :=
+  Matrix.toLin'.apply_symm_apply f
+#align matrix.to_lin'_to_matrix' Matrix.toLin'_toMatrix'
+
+@[simp]
+theorem LinearMap.toMatrix'_apply (f : (n → R) →ₗ[R] m → R) (i j) :
+    LinearMap.toMatrix' f i j = f (fun j' => if j' = j then 1 else 0) i := by
+  simp only [LinearMap.toMatrix', LinearEquiv.coe_mk, of_apply]
+  congr
```

Comment:
This congr isn't working and I have no idea why. The goal here is
```
↑f (↑(stdBasis R (fun x ↦ R) j) 1) i = ↑f (fun j' ↦ if j' = j then 1 else 0) i
```
`refine FunLike.congr_arg f ?_` doesn't work either